src/dtree.ml

```diff
@@ -0,0 +1,192 @@
+open Terms
+open Extra
+
+(** Type of the leaves of the tree.  See {!file:terms.ml}, {!recfield:rhs}. *)
+type action = (term_env, term) Bindlib.mbinder
+
+(** Tree type.  Leaves are the targets of the rewriting, i.e. when a leaf is
+    reached, the matched pattern is rewrote to the the leaf.  The
+    {!const:Node}[(i, t)] constructor allows to perform a switch among the
+    subtrees in [t] while the [i] option allows to choose on which term of the
+    stack the matching is to be done.  {!const:Fail} is a matching failure. *)
+type t = Leaf of action
+       | Node of node_data
+       | Fail
+
+(** Data contained in a node of the tree.  {!recfield:switch} contains the
+    term on which the switch that gave birth to this node has been performed
+    (or none if the node has been obtained by a default matrix);
+    {!recfield:swap} indicates whether the columns of the matrix have been
+    swapped before the switch, with an int indicating the column which has
+    been swapped with the first one (or None if no swap is performed) and
+    {!recfield:children} contains the subtrees. *)
+and node_data = { switch : term option
```

Comment:
Isn't the switch on just a symbol?

Reply:
It might be a lambda as well (in the future).

Reply:
Why not using a type Const of symb | Lambda?

Reply:
Thanks for adding an example of the tree one should get from compiling a simple rewrite system.

Reply:
Example and more doc in 7755e45 and 02ec495; tell me if it remains unclear.